chia/wallet/wallet_node.py

```diff
@@ -1377,52 +1379,77 @@ async def validate_block_inclusion(
+                    # Validate that the block is buried in the foliage by checking the signatures
+                    pk_m_sig.append(
+                        (
+                            en_block.reward_chain_block.proof_of_space.plot_public_key,
+                            en_block.foliage.foliage_block_data.get_hash(),
+                            en_block.foliage.foliage_block_data_signature,
+                        )
+                    )
+                    sigs_to_cache.append(en_block)
+
+                # This is the reward chain challenge. If this is in the cache, it means the prev block
+                # has been validated. We must at least check the first block to ensure they are connected
+                reward_chain_hash: bytes32 = en_block.reward_chain_block.reward_chain_ip_vdf.challenge
+                if idx != 0 and peer_request_cache.in_blocks_validated(reward_chain_hash):
+                    # As soon as we see a block we have already concluded is in the chain, we can quit.
+                    if idx > signatures_to_validate:
+                        break
                 else:
-                    next_block_rc_hash = reversed_blocks[idx + 1].reward_chain_block.get_hash()
-                    prev_hash = reversed_blocks[idx + 1].header_hash
+                    # Validate that the block is committed to by the weight proof
+                    if idx == 0:
+                        prev_block_rc_hash: bytes32 = block.reward_chain_block.get_hash()
+                        prev_hash = block.header_hash
+                    else:
+                        prev_block_rc_hash = blocks[idx - 1].reward_chain_block.get_hash()
+                        prev_hash = blocks[idx - 1].header_hash
 
-                if not en_block.prev_header_hash == prev_hash:
-                    self.log.error("Failed validation 5")
-                    return False
+                    if not en_block.prev_header_hash == prev_hash:
+                        self.log.error("Failed validation 5")
+                        return False
 
-                if len(en_block.finished_sub_slots) > 0:
-                    #  What to do here
-                    reversed_slots = en_block.finished_sub_slots.copy()
-                    reversed_slots.reverse()
-                    for slot_idx, slot in enumerate(reversed_slots[:-1]):
-                        hash_val = reversed_slots[slot_idx + 1].reward_chain.get_hash()
-                        if not hash_val == slot.reward_chain.end_of_slot_vdf.challenge:
-                            self.log.error("Failed validation 6")
+                    if len(en_block.finished_sub_slots) > 0:
+                        #  What to do here
+                        reversed_slots = en_block.finished_sub_slots.copy()
+                        reversed_slots.reverse()
+                        for slot_idx, slot in enumerate(reversed_slots[:-1]):
```

Comment:
FYI, not sure how big this list can be, but to avoid reversing the whole list, you could use a range 
```python
l = ["a", "b", "c", "d", "e"]
for idx in range(len(l)-2, -1, -1):
    slot = l[idx]
    prev_slot = l[idx+1]
    print(f"comparing {slot} with {prev_slot}")
```

Reply:
99.9% + of the time it will just be length 1.